include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,138 @@
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        return histogram.atomicAdd((tid / GroupSize) * GroupSize + shifted_tid, sdata.get(shifted_tid));
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+
+        for (; tid < KeyBucketCount; tid += GroupSize) {
+            sdata.set(tid, 0);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        uint32_t index = params.workGroupIndex * GroupSize * params.elementsPerWT + tid % GroupSize;
+        uint32_t endIndex = min(params.dataElementCount, index + GroupSize * params.elementsPerWT);
+
+        for (; index < endIndex; index += GroupSize)
+        {
+            uint32_t k = key.get(index);
+            if (robust && (k<params.minimum || k>params.maximum) )
+                continue;
+            sdata.atomicAdd(k - params.minimum, (uint32_t) 1);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+    }
+
+    void histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(HistogramAccessor) histogram, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        build_histogram(key, sdata, params);
+
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t histogram_value = sdata.get(tid);
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        uint32_t sum = inclusive_scan(histogram_value, sdata);
+        toroidal_histogram_add(tid, sum, histogram, sdata, params);
+
+        const bool is_last_wg_invocation = tid == (GroupSize - 1);
+        const uint16_t adjusted_key_bucket_count = ((KeyBucketCount - 1) / GroupSize + 1) * GroupSize;
+
+        for (tid += GroupSize; tid < adjusted_key_bucket_count; tid += GroupSize)
+        {
+            if (is_last_wg_invocation)
+            {
+                uint32_t startIndex = tid - tid % GroupSize;
+                sdata.set(startIndex, sdata.get(startIndex) + sum);
+            }
+
+            sum = inclusive_scan(sdata.get(tid), sdata);
+            toroidal_histogram_add(tid, sum, histogram, sdata, params);
+        }
+    }
+                
+    void scatter(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ValueAccessor) val, NBL_REF_ARG(HistogramAccessor) histogram, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        build_histogram(key, sdata, params);
+
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t shifted_tid = (tid + glsl::gl_SubgroupSize() * params.workGroupIndex) % GroupSize;
+
+        for (; shifted_tid < KeyBucketCount; shifted_tid += GroupSize)
+        {
+            uint32_t bucket_value = sdata.get(shifted_tid);
+            uint32_t exclusive_value = histogram.atomicSub(shifted_tid, bucket_value) - bucket_value;
+
+            sdata.set(shifted_tid, exclusive_value);
+        }
```

Comment:
ok this still makes sense to do with a toroidal shift in spite of https://github.com/Devsh-Graphics-Programming/Nabla/pull/690/files#r1621277804

but there's a different way to write this
```cpp
const uint32_t shift = glsl::gl_SubgroupSize()*params.workGroupIndex;
for (uint32_t shifted_tid=tid; vtid<KeyBucketCount; vtid+=GroupSize)
{
   // have to use modulo operator in case `KeyBucketCount<=2*GroupSize`, better hope KeyBucketCount is Power of Two
   const uint32_t shifted_tid = (vtid+shift)%KeyBucketCount;

   const uint32_t bucket_value = sdata.get(shifted_tid);
   const uint32_t firstOutputIndex = histogram.atomicSub(shifted_tid,bucket_value)-bucket_value;
   sdata.set(shifted_tid,firstOutputIndex);
}
```
doing a loop conditioned on `shifted_tid` would break unrolling during optimization, plus your `%GroupSize` wraps around differently than a `%KeyBucketCount` would (you get stuck within the [0,GroupSize) block, not really easing contetion).

Reply:
Done